File: 1844. Replace All Digits with Characters.rb
#* 1844. Replace All Digits with Characters
s = "a1c1e1"
# @param {String} s
# @return {String}
#* I DID A TON OF SOLUTIONS :)
#* SOL 1 | 2ND TRY
def replace_digits(s)
  shift = ->(c, x) { ((c.ord - 'a'.ord + x.to_i) % 26 + 'a'.ord).chr }
  s.chars.each_slice(2).map { |char, digit| [char, shift.call(char, digit)] }.flatten.join
end

puts replace_digits(s) 

#* SOL 2 | OG SOL CAME BACK
def replace_digits(s)
  res = []
  s.chars.each_slice(2) do |chr_a, chr_b|
      p res << chr_a
      p res << (chr_a.ord + chr_b.to_i).chr if chr_b
  end
  res.join
end
p replace_digits(s)
#* SOL 2.1 | READABLE/CONCISE
def replace_digits(s)
  s.chars.each_slice(2) do |a_of_pair, b_of_pair|
    s[b_of_pair] = (a_of_pair.ord + b_of_pair.to_i).chr if b_of_pair
  end
  s
end
#* SOL 2.2 | SUPER CONCISE
def replace_digits(s)
  s.chars.each_slice(2) {|a,b| s[b] = (a.ord + b.to_i).chr if b} and s
end
p replace_digits(s) # Output: "abcdef"

#* SOL 3
def replace_digits(s)
  s.chars.each_with_index{|chr,i| s[i] << (s[i-1].ord + chr.to_i).chr if i.odd?} and s
end
p replace_digits(s) # Output: "abcdef"


#* TEST
p x = s.chars.each_with_index.map {|c,i| c if i.even?}
p  y = s.chars.each_with_index.map {|c,i| c if i.odd?}
p x.zip(y)
p z = s.chars.each_with_index.partition {|ch, ind| ind.even?}

p a = s.chars.select.with_index.map {|c,i| c if i.even?}
p b = s.chars.select.with_index.map {|c,i| c if i.odd?}


  p c = s.chars.select.with_index { |c, i| i.even? }
  p d = s.chars.select.with_index { |c, i| i.odd? }
  c.zip(d)

  better = s.chars.each_slice(2).to_a
  p better

def replace_digits(s)
  p zipped = s.chars.each_slice(2).to_a
end
p replace_digits(s)
#*(END)[TEST]




#* lambda written out
  # shift = lambda do |c,x|
  #   ((c.ord - 'a'.ord + x.to_i) % 26 + 'a'.ord).chr 
  # end




#*WORKING . . . . .
#*
#* DOESNT WORK AGAIN
p 'DOESNT WORK OG CLOSE'
def replace_digits(s)
  i = 1
  s.chars.each_slice(2) do |char, digit|
    s[i] = ((char.ord - 'a'.ord + digit.to_i) % 26 + 'a'.ord).chr
    i += 2
  end
  s
end
p replace_digits(s)
p 'doesnt work' #* DOESNT WORK
def replace_digits(s)
  result = ""
  s.chars.each_slice(2) do |char, digit|
    result += ((char.ord - 'a'.ord + digit.to_i) % 26 + 'a'.ord).chr
  end
  result
end
p replace_digits(s)
#* . . . . .

#* MISC
p 7 % 1 == 0
p 7 % 2 == 0 

p 7 % 2 == 1 #* never saw this as odd check until today always % 1 == 0

File: 1844. Replace All Digits with Characters.js
s = "a1c1e1"
/**
 * @param {string} s
 * @return {string}
 */

//TESTS
console.log(s)
console.log(Array.from(s))
console.log(s.split(''))

var replaceDigits = function(s) {
  const x = Array.from(s).filter((char, index) => index % 2 === 0);
  const y = Array.from(s).filter((char, index) => index % 2 !== 0);
  return x.map((char, index) => [char, y[index]]);
}
console.log(replaceDigits(s)); // Output: [["a", "1"], ["c", "1"], ["e", "1"]]

function replaceDigits(s) {
  //return Array.from(s).reduce((acc, char, index) => { <= USE EITHER
  return s.split("").reduce((acc, char, index) => {
    if (index % 2 === 0) {
      acc.push([char, s[index + 1]]);
    }
    return acc;
  }, []);
}
console.log(replaceDigits(s))

function eachSlice(array, size) {
  const result = [];
  for (let i = 0; i < array.length; i += size) {
    result.push(array.slice(i, i + size));
  }
  return result;
}
console.log(eachSlice(s, 2)); // Output: [[1, 2], [3, 4], [5, 6]]

function eachSlice(s, size) {
  const result = [];
  const chars = Array.from(s);
  console.log
  for (let i = 0; i < chars.length; i += size) {
    result.push(chars.slice(i, i + size).join(""));
  }
  return result;
}

//console.log(eachSlice(s, 2)); // Output: ["a1", "c1", "e1"]




